### PR TITLE
refactor with tslint

### PR DESCRIPTION
**Pull request by Shinigami92**
- [x] I have read the [Contributor Guide](https://github.com/nteract/nteract/blob/master/CONTRIBUTING.md)

_Please dont merge this know_

Fix tslint errors (currently especially errors!)

Ignoring errors for rules:

- `class-name`
- `jsx-no-lambda-props`
- `jsx-no-lambda`
- `jsx-no-object-literal-props`
- `no-empty`
- `react-a11y-event-has-role`
- `react-component-classes-should-implement-scu`
- `react-pure-components-have-simple-attributes`

Also ignored some `ban-types` and `strict-type-predicates`

I will create a follow-up PR for the tslint rule `typedef`, this will boost the projects type-definitions

### CODE REVIEW

**Comment by Shinigami92** [comment]
@rgbkrk and/or @captainsafia you can start to review this baby
And I ...
Sleepy Cat

**Comment by rgbkrk** [comment]
Nice. I'll wrap up my PR and then review this in earnest. Have some good rest.

File: applications/desktop/__mocks__/enchannel-zmq-backend.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ class HokeySocket extends EventEmitter {
 }
 
 module.exports = {
-  createMainChannel: async function() {
+  async createMainChannel() {
```
Comment (rgbkrk) [inline applications/desktop/__mocks__/enchannel-zmq-backend.ts:15]: 👍

File: applications/desktop/__tests__/renderer/epics/close-notebook.spec.ts [hunk 2]
```diff
@@ -85,7 +87,9 @@ describe("closeNotebookEpic", () => {
     test("and allows aborting, triggering IPC close-notebook-canceled for app-wide quit orchestration", async () => {
       let registeredCallback;
       ipc.once = (event, callback) => {
-        if (event == "show-message-box-response") registeredCallback = callback;
+        if (event === "show-message-box-response") {
+          registeredCallback = callback;
+        }
       };
       ipc.send = (event, data) => {
         expect(event).toBe("show-message-box");
```
Comment (rgbkrk) [inline applications/desktop/__tests__/renderer/epics/close-notebook.spec.ts:95]: 👍

File: applications/desktop/__tests__/renderer/epics/kernel-launch.spec.ts [hunk 3]
```diff
@@ -53,7 +53,7 @@ describe("launchKernelEpic", () => {
     ]);
   });
 
-  test("calls launchKernelObservable if given the correct action", async function() {
+  test("calls launchKernelObservable if given the correct action", async () => {
     const action$ = ActionsObservable.of(
       actionsModule.launchKernel({
         kernelSpec: { spec: "hokey", name: "woohoo" },
```
Comment (rgbkrk) [inline applications/desktop/__tests__/renderer/epics/kernel-launch.spec.ts:59]: 👍

File: applications/desktop/__tests__/renderer/global-events.spec.ts [hunk 4]
```diff
@@ -133,7 +133,9 @@ describe("initGlobalHandlers", () => {
     const store = createStore(contentRef);
 
     ipc.on = event => {
-      if (event == "reload") done();
+      if (event === "reload") {
+        done();
+      }
     };
 
     globalEvents.initGlobalHandlers(contentRef, store);
```
Comment (rgbkrk) [inline applications/desktop/__tests__/renderer/global-events.spec.ts:141]: 👍

File: applications/desktop/src/notebook/menu.ts [hunk 5]
```diff
@@ -634,7 +636,7 @@ export function exportPDF(
           level: "success",
           action: {
             label: "Open PDF",
-            callback: function openPDF() {
+            callback() {
```
Comment (rgbkrk) [inline applications/desktop/src/notebook/menu.ts:639]: reasonable

File: applications/desktop/src/notebook/menu.ts [hunk 6]
```diff
@@ -666,7 +668,7 @@ export function storeToPDF(
   const state = store.getState();
   const notebookName = selectors.filepath(state, ownProps);
   const notificationSystem = state.app.get("notificationSystem");
-  if (notebookName == null) {
+  if (notebookName === null) {
```
Comment (rgbkrk) [inline applications/desktop/src/notebook/menu.ts:671]: I feel like these `==` checks are making sure we're not getting `null | undefined`.

**Review by rgbkrk — COMMENTED** [review]
Since I've had this review in draft I feel I should post some of it. Thank you for taking all the effort to get all our tslint and general code base cleaned up.

The scary thing about an ongoing big lint fixing PR, especially if WIP is that the changes on `master` will pile up, possibly negating prior reviews. There are a lot of changes in here that are super easy to merge that we might as well bring in and some that may require some more thinking where it should be together in a PR (`== null` checks becoming `=== null` is different behavior, since now `undefined` doesn't match).

File: applications/desktop/src/notebook/menu.ts (hunk 6 above)
Comment (Shinigami92) [reply applications/desktop/src/notebook/menu.ts:671]: Maybe I can write this to `!notebookName`
This would catch empty string also

File: applications/desktop/src/notebook/menu.ts (hunk 5 above)
Comment (Shinigami92) [reply applications/desktop/src/notebook/menu.ts:639]: I have some of this, I think in first run I can add a comment to this line with the old name 
The problem here is that this function name is only for human and the linter doesn't like human >:)

File: applications/desktop/__tests__/renderer/epics/close-notebook.spec.ts (hunk 2 above)
Comment (Shinigami92) [reply applications/desktop/__tests__/renderer/epics/close-notebook.spec.ts:95]: Checks like this dont alter the flow because it is checked against a static string

**Comment by Shinigami92** [comment]
When I'm back home I will rebase and ask if any PR should happens before 
Then we can merge this today or tomorrow 
Then I will start the next PR to improve the warnings 
Especialy at first the rule typedef

File: applications/desktop/src/notebook/native-window.ts [hunk 7]
```diff
@@ -67,7 +67,8 @@ export function setTitleFromAttributes(attributes: Attributes) {
     win.setTitle(title);
   } catch (e) {
     /* istanbul ignore next */
-    (function log1277() {
+    // log1277
```
Comment (Shinigami92) [inline applications/desktop/src/notebook/native-window.ts:70]: I added a comment here
In `openPDF` the callback can be associated by the action.label

File: applications/jupyter-extension/nteract_on_jupyter/app/components/last-saved.tsx [hunk 8]
```diff
@@ -58,7 +58,7 @@ class LastSaved extends React.PureComponent<LastSavedProps> {
   }
 
   render() {
-    if (this.props.date == null) {
+    if (this.props.date === null) {
```
Comment (Shinigami92) [inline applications/jupyter-extension/nteract_on_jupyter/app/components/last-saved.tsx:61]: Line 20 (`interface LastSavedProps`) says that date cant be `undefined`

File: package.json [hunk 9]
```diff
@@ -27,7 +27,7 @@
     "spawn:debug": "lerna run spawn:debug --scope nteract",
     "lint": "npm run lint:ts",
     "lint:fix": "npm run lint:ts:fix",
-    "lint:ts": "tslint --config tslint.json --project tsconfig.base.json",
+    "lint:ts": "tslint --config tslint.json --project tsconfig.base.json --force --format stylish",
```
Comment (Shinigami92) [inline package.json:30]: `--force` makes sure that tslint dont fail with an exit code != 0
`--format stylish` is a very nice output (colored) for humans

File: packages/data-explorer/src/VizControls.tsx [hunk 10]
```diff
@@ -116,7 +116,7 @@ const filterItem = (query: string, item: MenuItemType) => {
 };
 
 const getIcon = (title: string) => {
-  if (title === "X" || title === "Y" || title === "Size" || title == "Color") {
+  if (title === "X" || title === "Y" || title === "Size" || title === "Color") {
```
Comment (Shinigami92) [inline packages/data-explorer/src/VizControls.tsx:119]: Some lines like this can be rewritten to:
`["X", "Y", "Size", "Color"].includes(title)`
This is shorter and more dynamic

File: packages/data-explorer/src/VizControls.tsx [hunk 11]
```diff
@@ -235,23 +240,24 @@ const availableAreaTypes = [
...
-  chart: ChartOptions;
-  metrics: Array<{ name: string }>;
-  dimensions: Array<{ name: string }>;
-  updateChart: Function;
+  view: Dx.View;
+  chart: Dx.Chart;
+  metrics: Dx.Field[];
+  dimensions: Dx.Dimension[];
+  // TODO: leave "options: any" for now and improve typedef later
+  updateChart: (options: any) => void;
```
Comment (Shinigami92) [inline packages/data-explorer/src/VizControls.tsx:248]: Currently I have huge problems with this any
I hope to/Looking into fix this in my next `typedef`-PR

File: packages/data-explorer/src/charts/hierarchical.tsx [hunk 12]
```diff
@@ -58,8 +58,12 @@ const hierarchicalColor = (
   colorHash: { [index: string]: string },
   datapoint: Dx.Datapoint
 ) => {
-  if (datapoint.depth === 0) return "white";
-  if (datapoint.depth === 1) return colorHash[datapoint.key];
+  if (datapoint.depth === 0) {
```
Comment (Shinigami92) [inline packages/data-explorer/src/charts/hierarchical.tsx:61]: Can this be improved with a switch-case?
(Should be part of another PR)

File: packages/fs-observable/src/index.ts [hunk 13]
```diff
@@ -46,7 +46,7 @@ export const mkdirpObservable = bindNodeCallback(mkdirp);
  */
 export function readdirObservable(
   path: string,
-  options: { encoding?: string | null } | string | null
+  options?: { encoding?: string | null } | string | null
```
Comment (Shinigami92) [inline packages/fs-observable/src/index.ts:49]: This `?` is neede because a script wanted to call `readdirObservable` without providing the second param.
Because line 61 only checks against existing, it is handled like the `null`-case

File: packages/mathjax/src/provider.tsx [hunk 14]
```diff
@@ -53,17 +57,10 @@ class Provider extends React.Component<Props, State> {
...
-      return { ...state, input: props.input };
-    }
-    return null;
-  }
-
   componentDidMount() {
     const src = this.props.src;
 
-    if (src == null) {
+    if (src === undefined) {
```
Comment (Shinigami92) [inline packages/mathjax/src/provider.tsx:63]: Line 10 => `src?: string;`

File: packages/notebook-app-component/src/cell-creator.tsx [hunk 15]
```diff
@@ -144,7 +144,7 @@ class CellCreator extends React.PureComponent<ConnectedProps> {
       contentRef
     } = this.props;
 
-    if (id == null || typeof id != "string") {
+    if (id === undefined || typeof id !== "string") {
```
Comment (Shinigami92) [inline packages/notebook-app-component/src/cell-creator.tsx:147]: Line 36 => `id?: string;`

File: tslint.json [hunk 16]
```diff
@@ -1,23 +1,72 @@
...
     "tslint:recommended",
     "tslint-react",
     "tslint-react-a11y",
     "tslint-react-perf",
-    "tslint-config-prettier"
+    "tslint-config-prettier",
+    "tslint-eslint-rules"
   ],
-  "jsRules": {},
   "rules": {
```
Comment (Shinigami92) [inline tslint.json:11]: rules are ordered alphabetically

File: tslint.json [hunk 17]
```diff
@@ -1,23 +1,72 @@
 {
-  "defaultSeverity": "warn",
```
Comment (rgbkrk) [inline tslint.json:2]: Thanks for dropping some of the rules for now so we can iteratively get to a better state.

File: packages/presentational-components/src/components/header-editor.tsx [hunk 18]
```diff
@@ -241,7 +241,7 @@ export class HeaderEditor extends React.PureComponent<
               maxLength={280}
               maxLines={12}
               minLines={3}
-              multiline={true}
+              multiline
```
Comment (rgbkrk) [inline packages/presentational-components/src/components/header-editor.tsx:244]: ❤️

File: packages/fs-kernels/src/kernelspecs.ts [hunk 19]
```diff
@@ -51,7 +53,7 @@ function getKernelResources(kernelInfo: KernelInfo): Promise<KernelResource> {
       (data: string | Buffer) => ({
         files: files.map(x => path.join(kernelInfo.resourceDir, x)),
         name: kernelInfo.name,
-        resources_dir: kernelInfo.resourceDir, // eslint-disable-line camelcase
```
Comment (rgbkrk) [inline packages/fs-kernels/src/kernelspecs.ts:54]: Nice to see these little cleanups throughout.

File: tslint.json (hunk 17 above)
Comment (Shinigami92) [reply tslint.json:2]: defaultSeverity should stay on error, because this handles better the recommended rules
if any rule should have another severity it can be explicitly be set

**Comment by rgbkrk** [comment]
Alright let's bring this in and shake things out on master. We should hopefully be much closer to 100% tslinted so we can run it on CI.
